cpp/src/Ice/Properties.cpp

```diff
@@ -18,6 +18,102 @@ using namespace std;
 using namespace Ice;
 using namespace IceInternal;
 
+namespace
+{
+    /// Find a property in the Ice property set.
+    /// @param key The property name.
+    /// @param logWarnings Whether to log relevant warnings.
+    /// @return The property if found, nullopt otherwise.
+    optional<Property> findProperty(string_view key, bool logWarnings)
+    {
+        // Check if the property is legal.
+        LoggerPtr logger = getProcessLogger();
+        string::size_type dotPos = key.find('.');
+        if (dotPos != string::npos)
+        {
+            string_view prefix = key.substr(0, dotPos);
+            for (int i = 0; IceInternal::PropertyNames::validProps[i].properties != 0; ++i)
+            {
+                string pattern{IceInternal::PropertyNames::validProps[i].properties[0].pattern};
+
+                dotPos = pattern.find('.');
+
+                //
+                // Each top level prefix describes a non-empty
+                // namespace. Having a string without a prefix followed by a
+                // dot is an error.
+                //
+                assert(dotPos != string::npos);
+                string propPrefix = pattern.substr(0, dotPos);
+
+                // If the prefix doesn't match, continue to the next prefix.
+                if (IceUtilInternal::toUpper(propPrefix) != IceUtilInternal::toUpper(string{prefix}))
+                {
+                    continue;
+                }
+
+                for (int j = 0; j < IceInternal::PropertyNames::validProps[i].length; ++j)
+                {
+                    const IceInternal::Property& prop = IceInternal::PropertyNames::validProps[i].properties[j];
+
+                    if (IceUtilInternal::match(string{key}, prop.pattern))
+                    {
+                        if (prop.deprecated && logWarnings)
+                        {
+                            logger->warning("deprecated property: " + string{key});
+                        }
+                        return prop;
+                    }
+
+                    // Check for case-insensitive match.
+
+                    if (IceUtilInternal::match(
+                            IceUtilInternal::toUpper(string{key}),
+                            IceUtilInternal::toUpper(prop.pattern)))
+                    {
+                        if (logWarnings)
+                        {
+                            logger->warning(
+                                "unknown property: `" + string{key} + "'; did you mean `" + prop.pattern + "'");
```

Comment:
Can we use ostringstream?
```suggestion
                            ostringstream os;
                            os << "unknown property: `" << key << "'; did you mean `" << prop.pattern << "'";
                            logger->warning(os.str());
```